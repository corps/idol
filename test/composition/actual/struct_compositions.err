thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:345:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::panicking::panic
             at src/libcore/panicking.rs:49
   9: core::option::Option<T>::unwrap
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/macros.rs:12
  10: idol::type_composer::compose_types
             at src/type_composer.rs:142
  11: idol::type_resolver::TypeResolver::resolve_type
             at src/type_resolver.rs:70
  12: idol::type_resolver::resolve_types
             at src/type_resolver.rs:43
  13: idol::module_resolver::ModuleResolver::resolve
             at src/module_resolver.rs:63
  14: idol::module_resolver::resolve_module
             at src/module_resolver.rs:31
  15: idol::modules_store::ModulesStore::load
             at src/modules_store.rs:73
  16: idol::modules_store::ModulesStore::load_modules
             at src/modules_store.rs:57
  17: idol_composition::main
             at src/bin/idol_composition.rs:16
  18: std::rt::lang_start::{{closure}}
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/rt.rs:64
  19: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  21: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  22: std::rt::lang_start
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/rt.rs:64
  23: main
  24: __libc_start_main
  25: _start
