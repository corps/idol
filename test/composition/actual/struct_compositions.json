{"struct_compositions":{"module_name":"struct_compositions","types_by_name":{"NestedTwo":{"fields":{"two":{"field_name":"two","tags":[],"type_struct":{"literal":null,"primitive_type":"string","reference":{"module_name":"","qualified_name":"","type_name":""},"struct_kind":"Scalar"},"optional":false}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedTwo","type_name":"NestedTwo"},"options":[],"tags":[]},"AFieldOne":{"fields":{"a":{"field_name":"a","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedOne","type_name":"NestedOne"},"struct_kind":"Scalar"},"optional":false}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.AFieldOne","type_name":"AFieldOne"},"options":[],"tags":[]},"AFieldOneBFieldTwo":{"fields":{"a":{"field_name":"a","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedOne","type_name":"NestedOne"},"struct_kind":"Scalar"},"optional":false},"b":{"field_name":"b","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedTwo","type_name":"NestedTwo"},"struct_kind":"Scalar"},"optional":false}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.AFieldOneBFieldTwo","type_name":"AFieldOneBFieldTwo"},"options":[],"tags":[]},"AFieldOneOrBFieldTwo":{"fields":{"b":{"field_name":"b","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedTwo","type_name":"NestedTwo"},"struct_kind":"Scalar"},"optional":true},"a":{"field_name":"a","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedOne","type_name":"NestedOne"},"struct_kind":"Scalar"},"optional":true}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.AFieldOneOrBFieldTwo","type_name":"AFieldOneOrBFieldTwo"},"options":[],"tags":[]},"NestedOneTwo":{"fields":{"one":{"field_name":"one","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"","qualified_name":"","type_name":""},"struct_kind":"Scalar"},"optional":false},"two":{"field_name":"two","tags":[],"type_struct":{"literal":null,"primitive_type":"string","reference":{"module_name":"","qualified_name":"","type_name":""},"struct_kind":"Scalar"},"optional":false}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedOneTwo","type_name":"NestedOneTwo"},"options":[],"tags":[]},"NestedOne":{"fields":{"one":{"field_name":"one","tags":[],"type_struct":{"literal":null,"primitive_type":"int","reference":{"module_name":"","qualified_name":"","type_name":""},"struct_kind":"Scalar"},"optional":false}},"is_a":null,"named":{"module_name":"struct_compositions","qualified_name":"struct_compositions.NestedOne","type_name":"NestedOne"},"options":[],"tags":[]}},"types_dependency_ordering":["NestedTwo","NestedOne","NestedOneTwo","AFieldOne","AFieldOneOrBFieldTwo","AFieldOneBFieldTwo"]}}
