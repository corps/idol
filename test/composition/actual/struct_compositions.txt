=== module struct_compositions
struct_compositions.NestedTwo: {
  two: string # 
} # 
struct_compositions.NestedOne: {
  one: int # 
} # 
struct_compositions.NestedOneTwo: {
  one: int # 
  two: string # 
} # 
struct_compositions.AFieldOne: {
  a: struct_compositions.NestedOne # 
} # 
struct_compositions.AFieldOneOrBFieldTwo: {
  b: struct_compositions.NestedTwo? # 
  a: struct_compositions.NestedOne? # 
} # 
struct_compositions.AFieldOneBFieldTwo: {
  a: struct_compositions.NestedOne # 
  b: struct_compositions.NestedTwo # 
} # 
