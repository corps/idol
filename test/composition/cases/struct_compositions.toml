[NestedOne]
fields.one = "int"

[NestedTwo]
fields.two = "string"

[NestedOneTwo]
is_a = ["NestedOne", "NestedTwo"]

[AFieldOne]
fields.a = ["NestedOne"]

[AFieldOneBFieldTwo]
is_a = "AFieldOne"
fields.b = "NestedTwo"

[AFieldOneOrBFieldTwo]
is_a = "AFieldOne"
variance = "Contravariant"
fields.b = "NestedTwo"

