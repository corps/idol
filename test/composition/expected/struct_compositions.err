Error processing modules: While resolving type SystemPerson in module struct_compositions: local type Person does not exist
Error: 1
