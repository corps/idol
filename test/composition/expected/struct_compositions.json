{
  "struct_compositions": {
    "module_name": "struct_compositions",
    "types_by_name": {
      "AFieldOne": {
        "fields": {
          "a": {
            "field_name": "a",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "struct_compositions",
                "qualified_name": "struct_compositions.NestedOne",
                "type_name": "NestedOne"
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.AFieldOne",
          "type_name": "AFieldOne"
        },
        "options": [],
        "tags": []
      },
      "AFieldOneBFieldTwo": {
        "fields": {
          "a": {
            "field_name": "a",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "struct_compositions",
                "qualified_name": "struct_compositions.NestedOne",
                "type_name": "NestedOne"
              },
              "struct_kind": "Scalar"
            }
          },
          "b": {
            "field_name": "b",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "struct_compositions",
                "qualified_name": "struct_compositions.NestedTwo",
                "type_name": "NestedTwo"
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.AFieldOneBFieldTwo",
          "type_name": "AFieldOneBFieldTwo"
        },
        "options": [],
        "tags": []
      },
      "AFieldOneOrBFieldTwo": {
        "fields": {
          "a": {
            "field_name": "a",
            "optional": true,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "struct_compositions",
                "qualified_name": "struct_compositions.NestedOne",
                "type_name": "NestedOne"
              },
              "struct_kind": "Scalar"
            }
          },
          "b": {
            "field_name": "b",
            "optional": true,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "struct_compositions",
                "qualified_name": "struct_compositions.NestedTwo",
                "type_name": "NestedTwo"
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.AFieldOneOrBFieldTwo",
          "type_name": "AFieldOneOrBFieldTwo"
        },
        "options": [],
        "tags": []
      },
      "NestedOne": {
        "fields": {
          "one": {
            "field_name": "one",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "",
                "qualified_name": "",
                "type_name": ""
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.NestedOne",
          "type_name": "NestedOne"
        },
        "options": [],
        "tags": []
      },
      "NestedOneTwo": {
        "fields": {
          "one": {
            "field_name": "one",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "int",
              "reference": {
                "module_name": "",
                "qualified_name": "",
                "type_name": ""
              },
              "struct_kind": "Scalar"
            }
          },
          "two": {
            "field_name": "two",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "string",
              "reference": {
                "module_name": "",
                "qualified_name": "",
                "type_name": ""
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.NestedOneTwo",
          "type_name": "NestedOneTwo"
        },
        "options": [],
        "tags": []
      },
      "NestedTwo": {
        "fields": {
          "two": {
            "field_name": "two",
            "optional": false,
            "tags": [],
            "type_struct": {
              "literal": null,
              "primitive_type": "string",
              "reference": {
                "module_name": "",
                "qualified_name": "",
                "type_name": ""
              },
              "struct_kind": "Scalar"
            }
          }
        },
        "is_a": null,
        "named": {
          "module_name": "struct_compositions",
          "qualified_name": "struct_compositions.NestedTwo",
          "type_name": "NestedTwo"
        },
        "options": [],
        "tags": []
      }
    },
    "types_dependency_ordering": [
      "NestedTwo",
      "NestedOne",
      "NestedOneTwo",
      "AFieldOne",
      "AFieldOneOrBFieldTwo",
      "AFieldOneBFieldTwo"
    ]
  }
}
